src/ObjWriting/XModel/Export/XModelBinWriter.cpp

```diff
@@ -0,0 +1,474 @@
+     *       2           = Number of objects
+     *       0x87D4      = Start of object hash
+     *       0           = Object index
+     *       Test_Obj    = Object name
+     *       0x87D4      = Start of object hash
+     *       1           = Object index
+     *       Test_Obj2   = Object name
+     */
+    enum XModelBinHash : int32_t
+    {
+        COMMENT = 0xC355,
+        MODEL = 0x46C8,
+        VERSION = 0x24D1,
+
+        BONE_COUNT = 0x76BA,
+        BONE = 0xF099,
+        BONE_INDEX = 0xDD9A,
+        OFFSET = 0x9383, // Bone/Vertex Offset
+        BONE_SCALE = 0x1C56,
+        BONE_MATRIX_X = 0xDCFD,
+        BONE_MATRIX_Y = 0xCCDC,
+        BONE_MATRIX_Z = 0xFCBF,
+
+        VERT16_COUNT = 0x950D,
+        VERT32_COUNT = 0x2AEC,
+        VERT16 = 0x8F03,
+        VERT32 = 0xB097,
+        VERT_WEIGHT_COUNT = 0xEA46,
+        VERT_WEIGHT = 0xF1AB,
+
+        FACE_COUNT = 0xBE92,
+        TRIANGLE16 = 0x6711,
+        TRIANGLE32 = 0x562F,
+
+        NORMAL = 0x89EC,
+        COLOR = 0x6DD8,
+        UV = 0x1AD4,
+
+        OBJECT_COUNT = 0x62AF,
+        OBJECT = 0x87D4,
+
+        MATERIAL_COUNT = 0xA1B2,
+        MATERIAL = 0xA700,
+        MATERIAL_TRANSPARENCY = 0x6DAB,
+        MATERIAL_AMBIENT_COLOR = 0x37FF,
+        MATERIAL_INCANDESCENCE = 0x4265,
+        MATERIAL_COEFFS = 0xC835,
+        MATERIAL_GLOW = 0xFE0C,
+        MATERIAL_REFRACTIVE = 0x7E24,
+        MATERIAL_SPECULAR_COLOR = 0x317C,
+        MATERIAL_REFLECTIVE_COLOR = 0xE593,
+        MATERIAL_REFLECTIVE = 0x7D76,
+        MATERIAL_BLINN = 0x83C7,
+        MATERIAL_PHONG = 0x5CD2
+    };
+
+protected:
+    void PrepareVertexMerger(const XModelCommon& xmodel)
+    {
+        m_vertex_merger = VertexMerger(xmodel.m_vertices.size());
+
+        auto vertexOffset = 0u;
+        for (const auto& vertex : xmodel.m_vertices)
+        {
+            XModelVertexBoneWeights weights{0, 0};
+
+            if (vertexOffset < xmodel.m_vertex_bone_weights.size())
+                weights = xmodel.m_vertex_bone_weights[vertexOffset];
+
+            m_vertex_merger.Add(VertexMergerPos{vertex.coordinates[0],
+                                                vertex.coordinates[1],
+                                                vertex.coordinates[2],
+                                                xmodel.m_bone_weight_data.weights.empty() ? nullptr : &xmodel.m_bone_weight_data.weights[weights.weightOffset],
+                                                weights.weightCount});
+
+            vertexOffset++;
+        }
+    }
+
+    void WriteAlignedString(const std::string& string)
+    {
+        const auto paddingSize = ((string.size() + 1 + 0x3) & 0xFFFFFFFFFFFFFC) - (string.size() + 1);
+        const auto padding = std::make_unique<uint8_t[]>(static_cast<size_t>(paddingSize));
+
+        m_writer.WriteNullTerminatedString(string);
+        m_writer.Write(reinterpret_cast<uint8_t*>(padding.get()), static_cast<uint32_t>(paddingSize));
```

Comment:
What you are doing is you are writing a zero terminated string with padding to the next multiple of 4.
You are allocating new uninitialized memory for this and write it directly without initialization?

That seems slow and may leak previously allocated memory values.
Padding can at max be 3 bytes long.

```cpp
static constexpr uint8_t PADDING[4] = {0u, 0u, 0u, 0u};
const auto paddingSize = (4u - (string.size() + 1u) % 4u) % 4u;

m_writer.WriteNullTerminatedString(string);
m_writer.Write(PADDING, paddingSize);
```

(Btw there is a utility method `Utils/Alignment.h => Align` for aligning to next multiple of in case you need it)